### PR TITLE
Fix Search All Open Graphs in New Find View 

### PR DESCRIPTION
**Pull request by Delphinus8821**
### Prerequisites

- [X] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change
The new find view added an improvement to be able to search all graphs. This fix allows for the active graph to swap to where the found element is. This works for "Find All", "Find Next" and "Find Previous" on the Basic Find View, Find and Replace and the Advanced Find. 

### Why Should This Be In Core?
Easily swaps between graphs to show the element where the text has been found. 

### Benefits
Allows for users to easily find and replace elements across multiple graphs. 

### Verification Process
1. Open multiple graphs with various text strings 
2. Use the basic find to "find next" a few times to go between graphs
3. Repeat with find and replace and the advanced find

### Applicable Issues
#1696

### CODE REVIEW

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/FindViewUtilities.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        final Set<TopComponent> topComponents = WindowManager.getDefault().getRegistry().getOpened();
+        if (topComponents != null) {
+            for (final TopComponent component : topComponents) {
+                if ((component instanceof VisualGraphTopComponent) && ((VisualGraphTopComponent) component).getGraphNode().getGraph().getId().equals(graph.getId())) {
+                    EventQueue.invokeLater(new Runnable() {
+                        @Override
+                        public void run() {
+                            ((VisualGraphTopComponent) component).requestActive();
+                        }
+                    });
```
Comment (OrionsGuardian) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/FindViewUtilities.java:51]: Can we put in a "break;" after the invokeLater() call? 
Shouldn't need to keep iterating through the for loop.

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]

**Comment by mimosa2** [comment]
Thanks @Delphinus8821. The search works with opened graphs. A minor issue is when no results found and I tried FindNext option to search the graphs and got this error. Is it possible to avoid this by displaying a "No results Found" message?     
FindView_SearchAllOpenGraphs

**Comment by Delphinus8821** [comment]
> Thanks @Delphinus8821. The search works with opened graphs. A minor issue is when no results found and I tried FindNext option to search the graphs and got this error. Is it possible to avoid this by displaying a "No results Found" message? !

I have been unable to reproduce this error (it looks like it is being caused by the attribute editor) but I will look to add something for when there are no results found.

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/BasicFindGraphSelectionPlugin.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    public BasicFindGraphSelectionPlugin(final BasicFindReplaceParameters parameters, final boolean selectAll) {
+        this.elementType = parameters.getGraphElement();
+        this.selectAll = selectAll;
+        this.removeFromCurrentSelection = parameters.isRemoveFrom();
+        this.searchAllGraphs = parameters.isSearchAllGraphs();
+    }
+
+    @Override
+    protected void edit(GraphWriteMethods graph, PluginInteraction interaction, PluginParameters parameters) throws InterruptedException, PluginException {
```
Comment (antares1470) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/BasicFindGraphSelectionPlugin.java:47]: parameters final

File: CoreFindView/test/unit/src/au/gov/asd/tac/constellation/views/find2/FindViewControllerNGTest.java [hunk 3]
```diff
@@ -258,50 +260,13 @@ public void testRetriveMatchingElements() {
...
+                basicFindPlugin = new BasicFindPlugin(instance.getCurrentBasicFindParameters(), true, false);
+                PluginExecution.withPlugin(basicFindPlugin).executeLater(graph);
                 verify(pluginExecution, times(2)).executeLater(Mockito.eq(graph));
-                verify(pluginExecution).executeLater(Mockito.eq(graph2));
-
             }
         }
     }
 
-//            // Populate the attribute list with the attributes of the given type
```
Comment (antares1470) [inline CoreFindView/test/unit/src/au/gov/asd/tac/constellation/views/find2/FindViewControllerNGTest.java:269]: just curious as to the reason for this whole code block being commented out?

**Review by antares1470 — CHANGES_REQUESTED** [review]
Looks to be working in testing

File: CoreFindView/test/unit/src/au/gov/asd/tac/constellation/views/find2/FindViewControllerNGTest.java (hunk 3 above)
Comment (Delphinus8821) [reply CoreFindView/test/unit/src/au/gov/asd/tac/constellation/views/find2/FindViewControllerNGTest.java:269]: Not really sure why, it was like that before I looked at it

Comment (antares1470) [reply CoreFindView/test/unit/src/au/gov/asd/tac/constellation/views/find2/FindViewControllerNGTest.java:269]: ah right, I just realised you were removing that code. All good

**Review by antares1470 — APPROVED** [review]

**Comment by mimosa2** [comment]
Find and Replace work perfect. I can toggle between graphs using Find Next and Find Previous w/o any issues and the Results Found reflects the hits correctly.

**Review by mimosa2 — APPROVED** [review]

**Review by antares1470 — CHANGES_REQUESTED** [review]
Was just doing some testing with the advanced find. Looks like if works as intended if you have "Search all open graphs" ticked. If it's not ticked though, it gets the results of the current graph, but then if you switch graphs and run the find, it accumulates the results rather than replace so when you "Find Next" (or Previous), it alternates between results in each graph (without switching graphs)
advancedFindBug

**Review by arcturus2 — CHANGES_REQUESTED** [review]
@Delphinus8821 I've found an issue after closing graphs. The search all graphs feature still thinks that the closed graphs exist.

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/ActiveFindResultsList.java [hunk 4]
```diff
@@ -102,4 +105,23 @@ public static void addToAdvancedFindResultsList(final FindResultsList additions)
...
     }
+
+    /**
+     * Gets how many unique graphs are found in the list - will equal the number of graphs open when the search was done
+     *
+     * @param findResultsList
+     * @return
+     */
+    public static int getUniqueGraphCount(final FindResultsList findResultsList) {
+        List<String> uniqueGraphs = new ArrayList();
```
Comment (antares1470) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/ActiveFindResultsList.java:116]: final

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/ActiveFindResultsList.java [hunk 5]
```diff
@@ -102,4 +105,23 @@ public static void addToAdvancedFindResultsList(final FindResultsList additions)
...
+        if (findResultsList != null) {
+            for (final FindResult result : findResultsList) {
+                String graphId = result.getGraphId();
```
Comment (antares1470) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/ActiveFindResultsList.java:119]: final

File: CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/ActiveFindResultsList.java [hunk 6]
```diff
@@ -102,4 +105,23 @@ public static void addToAdvancedFindResultsList(final FindResultsList additions)
...
+                if (!uniqueGraphs.contains(graphId)) {
+                    uniqueGraphs.add(graphId);
+                }
```
Comment (antares1470) [inline CoreFindView/src/au/gov/asd/tac/constellation/views/find2/utilities/ActiveFindResultsList.java:122]: Given you are just returning the list size and you're wanting the elements in the list to be unique, could possibly use a Set here instead (would save doing the contains check)

**Review by antares1470 — CHANGES_REQUESTED** [review]

**Review by antares1470 — APPROVED** [review]

**Pull request merged**